editor/plugins/sprite_frames_editor_plugin.cpp

```diff
@@ -503,6 +503,73 @@ void SpriteFramesEditor::_update_show_settings() {
 		toggle_settings_button->set_icon(get_editor_theme_icon(split_sheet_settings_vb->is_visible() ? SNAME("Forward") : SNAME("Back")));
 	}
 }
+bool SpriteFramesEditor::_matches_background_color(const Color &background_color, const Color &pixel_color) {
+	return background_color.is_equal_approx(pixel_color) || (background_color.a == 0 && pixel_color.a == 0);
+}
+
+Size2i SpriteFramesEditor::_estimate_sprite_sheet_size(const Ref<Texture2D> texture) {
+	Size2i default_size = Size2i(4, 4);
+	Ref<Image> image = texture->get_image();
+	Size2i size = texture->get_size();
+
+	Color assumed_background_color = image->get_pixel(0, 0);
+
+	bool contiguous_lines = true;
+	int division_count = 1;
+	for (int x = 0; x < size.x; x++) {
+		bool background_vert_line = true;
+		for (int y = 0; y < size.y; y++) {
+			if (!_matches_background_color(assumed_background_color, image->get_pixel(x, y))) {
+				background_vert_line = false;
+			}
+		}
+		if (contiguous_lines) {
+			if (!background_vert_line) {
+				contiguous_lines = false;
+			}
+		} else {
+			if (background_vert_line) {
+				contiguous_lines = true;
+				division_count++;
+			}
+		}
+	}
+	if (contiguous_lines && division_count > 1) {
+		division_count--;
+	}
+	default_size.x = division_count;
+
+	contiguous_lines = true;
+	division_count = 1;
+	for (int y = 0; y < size.y; y++) {
+		bool background_horiz_line = true;
+		for (int x = 0; x < size.x; x++) {
+			if (!_matches_background_color(assumed_background_color, image->get_pixel(x, y))) {
+				background_horiz_line = false;
+			}
+		}
+		if (contiguous_lines) {
+			if (!background_horiz_line) {
+				contiguous_lines = false;
+			}
+		} else {
+			if (background_horiz_line) {
+				contiguous_lines = true;
+				division_count++;
+			}
+		}
+	}
+	if (contiguous_lines && division_count > 1) {
+		division_count--;
+	}
+	default_size.y = division_count;
+
+	if (default_size.x == 1 && default_size.y == 1) {
+		default_size.x = 4;
+		default_size.y = 4;
+	}
+	return default_size;
+}
```

Comment:
A little simplified/shortened. AFAICT should do pretty much same thing (haven't tested :upside_down_face:). Also should be faster as you do not early-`break` out of the inner loops.
```suggestion
Size2i SpriteFramesEditor::_estimate_sprite_sheet_size(const Ref<Texture2D> texture) {
	Ref<Image> image = texture->get_image();
	Size2i size = texture->get_size();

	Color assumed_background_color = image->get_pixel(0, 0);
	Size2i sheet_size = Size2i(0, 0);

	bool previous_line_contigous = true;
	for (int x = 0; x < size.x; x++) {
		int y = 0;
		while (y < size.y && _matches_background_color(assumed_background_color, image->get_pixel(x, y))) {
			y++;
		}
		bool current_line_contigous = (y == size.y);
		if (previous_line_contigous && !current_line_contigous) {
			sheet_size.x++;
		}
		previous_line_contigous = current_line_contigous;
	}

	previous_line_contigous = true;
	for (int y = 0; y < size.y; y++) {
		int x = 0;
		while (x < size.x && _matches_background_color(assumed_background_color, image->get_pixel(x, y))) {
			x++;
		}
		bool current_line_contigous = (x == size.x);
		if (previous_line_contigous && !current_line_contigous) {
			sheet_size.y++;
		}
		previous_line_contigous = current_line_contigous;
	}
	
	if (sheet_size.x == 0) {
		sheet_size.x = 4;
	}
	if (sheet_size.y == 0) {
		sheet_size.y = 4;
	}

	return sheet_size;
}
```

Reply:
Thank you for the code submission Kleonc. I see how your code clarifies things. The key difference is you add a row/column when it hits a sprite, while I was doing it when it was hitting the background.

I've tested it quite a bit, and made my own changes and cleanups to it (changing how it reverts to default size was my biggest thing, as well as ditching "contiguous" as a variable name). I think it's definitely in a better state, and you're totally right with optimising it to break out of a loop when a non-background pixel is hit.

Thank you very much!